### PR TITLE
thanks nows read data from the packages and json-api (even tho the data is missing)

### PR DESCRIPTION
**Pull request by tomdottom**
@phildini I think this is most of the way to addressing #7 albeit a bit rough around the edges.

Of note:
- project no longer contains the mock db `thanks.json`
- locally we read data from either `PKG-INFO` for eggs or `METADATA` for dists
  - theoretically we could also parse these for patreon of paypal links
- added a message to get users to submit PRs asking authors/maintainers to funding metadata
  - this probably needs reworking

I've yet to find a project on PyPI which contains a funding url so, although I believe it all works, I haven't been able to confirm against the live wharehouse.

### CODE REVIEW

**Comment by phildini** [comment]
`thanks` itself has funding data: https://pypi.org/project/thanks/

Hoping to look at this PR in more depth later today.

**Comment by tomdottom** [comment]
True, but even `thanks` doesn't have it's _funding_ url on pypi yet :)
Do you know of any other projects which already have it?

**Comment by phildini** [comment]
Eh? I see the funding URL listed on the pypi page.

screen shot 2018-05-21 at 11 54 29 am

**Comment by tomdottom** [comment]
:man_facepalming: I guess I was a fool to assume it would appear in the json-api results.

Any idea where the code base which serves the json-api lives?

However, it does extract the url from local installs.

File: thanks/package_tools.py [hunk 1]
```diff
@@ -15,31 +19,94 @@ def get_local_dist(package_name):
...
-        metadata = json.load(fh)
+        metadata = parse_metadata(fh.read())
     return metadata
 
 
 def get_funding_data(metadata):
-        return (
-            metadata['extensions']['python.details']['project_urls']['Funding']
-        )
+    print(metadata)
```
Comment (phildini) [inline thanks/package_tools.py]: Is this leftover debugging code?

File: thanks/package_tools.py [hunk 2]
```diff
@@ -15,31 +19,94 @@ def get_local_dist(package_name):
...
+    return metadata.get('funding_url')
 
 
-def get_local_funding_metadata(package_name):
+def get_local_dist_metadata_filepath(dist):
+    # Dist filename syntax
+    # name ["-" version ["-py" pyver ["-" required_platform]]] "." ext
+    # https://setuptools.readthedocs.io/en/latest/formats.html#filename-embedded-metadata
+    filename = ''.join(chain(*takewhile(lambda x: x[1], (
```
Comment (phildini) [inline thanks/package_tools.py]: This looks cool as hell; could you add a comment explaining what this is doing? These are parts of itertools I'm not familiar with

**Review by phildini — COMMENTED** [review]
Overall I think this looks good, minus some nits.

Also, I think tests are failing because ddt needs to be added to `test_requires` or in some other way referenced in setup.py

File: thanks/package_tools.py (hunk 1 above)
Comment (tomdottom) [reply thanks/package_tools.py]: :man_facepalming:

File: thanks/package_tools.py (hunk 2 above)
Comment (tomdottom) [reply thanks/package_tools.py]: :thinking: if it needs a comment maybe I'm trying to be too clever.

Basically it takes all components parts of the filename until it finds a `None` then concatenates them into the finished filename.

Will refactor to try and make it self-explanatory and cognitively easy on the reader (which will probably be me sometime in the future)
git blame

**Comment by tomdottom** [comment]
Quick search seems to indicate that we'd have to expose the funding url in the json-api [here](https://github.com/pypa/warehouse/blob/40fbc32fef7c7ffe41cd18f3f8951578555db1aa/warehouse/legacy/api/json.py#L162)

I was about to suggest looking in the xmlrpc data but then encountered this:

```
Fault: <Fault -32500: 'RuntimeError: This API has been deprecated. Use https://pypi.org/thanks/0.0.8/json instead. The XMLRPC method release_data can be used in the interim, but will be deprecated in the future.'>
```

**Comment by tomdottom** [comment]
Just found https://github.com/pypa/warehouse/issues/3831 which will probably work for us once it's completed

File: setup.py [hunk 3]
```diff
@@ -60,7 +62,7 @@
         'Programming Language :: Python :: 3.5',
         'Programming Language :: Python :: 3.6',
     ],
-    test_suite='tests',
+    test_suite='nose.collector',
```
Comment (phildini) [inline setup.py:65]: Could you talk a bit about why we're moving to `nose`?

File: thanks/package_tools.py [hunk 4]
```diff
@@ -15,31 +18,99 @@ def get_local_dist(package_name):
...
+    filename = '{}.{}'.format(filename, ext)
+    path = os.path.join(dist.location, filename, metadata_file)
+
+    if ext:
+        return path
+    else:
+        return None
+
+
+metadata_patterns = re.compile(r"""
```
Comment (phildini) [inline thanks/package_tools.py:70]: This is a cool little collection of regexes, nice.

File: thanks/thanks.py [hunk 5]
```diff
@@ -61,15 +53,29 @@ def _display_thanks(self):
                 horizontal_bar=' ',
                 vertical_bar=' ',
             ))
+            cprint(
+                ''.join([
+                    "If you see projects with ",
```
Comment (phildini) [inline thanks/thanks.py:58]: This notice is lovely, and nicely viral in growing the packages which will have funding info.

**Review by phildini — COMMENTED** [review]
Overall this looks great. I have a question about using `nose`, but once we've chatted about that, ready to approve.

File: setup.py (hunk 3 above)
Comment (tomdottom) [reply setup.py:65]: There was a small note in the commit.

Basically `python setup.py test` was swallowing some simple errors (such as that the requests package wasn't installed.

`nose` presented these errors clearly and reminded me of my obvious error.

**Review by phildini — APPROVED** [review]

**Pull request merged**